File: python-worker/flags.py
import os
from pathlib import Path
import yaml

_flags = {}


def load_flags(path: str = "config/flags.yaml") -> None:
    global _flags
    config_path = Path(path)
    if not config_path.exists():
        _flags = {}
        return
    with config_path.open('r', encoding='utf-8') as handle:
        data = yaml.safe_load(handle) or {}
    items = (data.get('flags') or {})
    _flags = {key: bool(value.get('default', False)) for key, value in items.items()}


def is_enabled(name: str) -> bool:
    env_key = f"FLAG_{name.upper()}"
    if env_key in os.environ:
        return os.environ[env_key].lower() == 'true'
    return _flags.get(name, False)


load_flags()

File: backend/src/flags.ts
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';

type FlagMap = Record<string, boolean>;
let flags: FlagMap = {};

const DEFAULT_LOCATIONS = [
  path.resolve(process.cwd(), 'config/flags.yaml'),
  path.resolve(__dirname, '../../config/flags.yaml'),
  path.resolve(__dirname, '../config/flags.yaml'),
];

export function loadFlags(filePath?: string) {
  const candidatePaths = filePath ? [filePath] : DEFAULT_LOCATIONS;
  const resolvedPath = candidatePaths.find((candidate) => fs.existsSync(candidate));

  if (!resolvedPath) {
    flags = {};
    return;
  }

  const contents = fs.readFileSync(resolvedPath, 'utf8');
  const doc = yaml.load(contents) as any;
  const items = (doc?.flags ?? {}) as Record<string, { default?: boolean }>;
  flags = Object.fromEntries(
    Object.entries(items).map(([key, value]) => [key, Boolean(value?.default)])
  );
}

export function isEnabled(name: string): boolean {
  const envKey = `FLAG_${name.toUpperCase()}`;
  if (process.env[envKey] !== undefined) {
    return process.env[envKey]!.toLowerCase() === 'true';
  }
  return flags[name] ?? false;
}

loadFlags();
